### PR TITLE
Untrack previously installed submgr dependencies

### PR DESCRIPTION
**Pull request by r0x0d**
After performing a rollback on the system, the python3-syspurpose package is removed from the system causing some packages to report an error when `yum check` is executed after the rollback.

Jira reference: https://issues.redhat.com/browse/OAMG-5749
Bugzilla reference (If any):

### CODE REVIEW

File: convert2rhel/subscription.py [hunk 1]
```diff
@@ -247,8 +247,15 @@ def install_rhel_subscription_manager():
     else:
         loggerinst.info("Packages installed:\n%s" % "\n".join(rpms_to_install))
         pkg_names = get_installed_submgr_pkg_names(rpms_to_install)
-        utils.changed_pkgs_control.track_installed_pkgs(pkg_names)
+        if "python3-syspurpose" in pkg_names:
+            loggerinst.debug(
+                "Untracking python3-syspurpose from installed packages"
+                "as it could be a system package installed before this process."
+            )
+            pkg_names.remove("python3-syspurpose")
```
Comment (r0x0d) [inline convert2rhel/subscription.py:255]: I'm wondering, should we check beforehand if the packages to remove are already installed on the system? Maybe with `rpm -q python3-syspurpose`? 

If we find out that the mentioned package is already present in the system before installing `subscription-manager`, what should we do? Leave it as it is? Proceed with the reinstallation of the package for the correct version for `subscription-manager`?

File: convert2rhel/subscription.py [hunk 2]
```diff
@@ -242,13 +246,33 @@ def install_rhel_subscription_manager():
         # When using the original system repos, we need YUM/DNF to expand the $releasever by itself
         set_releasever=False,
     )
-    if ret_code:
-        loggerinst.critical("Failed to install subscription-manager packages. See the above yum output for details.")
-    else:
```
Comment (SpyTec) [inline convert2rhel/subscription.py:247]: Just a nitpick. We could keep `if ret_code:` and simply remove the else statement since critical is a SysExit

File: convert2rhel/subscription.py [hunk 3]
```diff
@@ -223,13 +223,17 @@ def remove_original_subscription_manager():
 
 
 def install_rhel_subscription_manager():
-    loggerinst.info("Installing subscription-manager RPMs.")
+    loggerinst.info("Checking for subscription-manager RPMs.")
```
Comment (SpyTec) [inline convert2rhel/subscription.py:226]: We're still installing subscription-manager in the method. We just have more checks to track and untrack the now installed packages

File: convert2rhel/subscription.py [hunk 4]
```diff
@@ -242,13 +246,33 @@ def install_rhel_subscription_manager():
...
-    else:
-        loggerinst.info("Packages installed:\n%s" % "\n".join(rpms_to_install))
-        pkg_names = get_installed_submgr_pkg_names(rpms_to_install)
-        utils.changed_pkgs_control.track_installed_pkgs(pkg_names)
+    if not ret_code:
+        loggerinst.info("\nPackages installed:\n%s" % "\n".join(rpms_to_install))
+
+        # Remove any packages that are present in the system before the install
+        [pkg_names.remove(rpm) for rpm in rpms_to_untrack if rpm in pkg_names]
+        loggerinst.debug("Untracked pacakges: %r" % rpms_to_untrack)
```
Comment (SpyTec) [inline convert2rhel/subscription.py:254]:
```suggestion
        loggerinst.debug("Untracked packages: %r" % rpms_to_untrack)
```

File: convert2rhel/subscription.py [hunk 5]
```diff
@@ -242,13 +246,33 @@ def install_rhel_subscription_manager():
...
         loggerinst.debug("Tracking installed packages: %r" % pkg_names)
+        utils.changed_pkgs_control.track_installed_pkgs(pkg_names)
+    else:
+        loggerinst.critical("Failed to install subscription-manager packages. See the above yum output for details.")
+
+
+def check_for_pkg_presence(pkg_names):
+    """Returns a list of packages that are already present in the system before
+    installing any packages needed by subscrpition-manager
```
Comment (SpyTec) [inline convert2rhel/subscription.py:263]:
```suggestion
    """Check that packages are installed on the system.
```

File: convert2rhel/subscription.py [hunk 6]
```diff
@@ -242,13 +246,33 @@ def install_rhel_subscription_manager():
...
+
+    :param pkg_names: List of package names
+    """
+    rpms_present = []
+    # Check for already installed packages that subscription-manager uses.
+    # If found any, then add it to a list an consider it as an untracked pkg.
```
Comment (SpyTec) [inline convert2rhel/subscription.py:269]:
```suggestion
    # Check for already installed packages
    # If a package is installed, add it to a list which is returned
```

File: convert2rhel/unit_tests/subscription_test.py [hunk 7]
```diff
@@ -293,12 +293,43 @@ def test_remove_original_subscription_manager(self):
...
+    @unit_tests.mock(pkghandler, "call_yum_cmd", lambda a, b, enable_repos, disable_repos, set_releasever: (None, 0))
+    @unit_tests.mock(logging.Logger, "info", LogMocked())
     @unit_tests.mock(os.path, "isdir", lambda x: True)
     @unit_tests.mock(os, "listdir", lambda x: ["filename"])
+    @unit_tests.mock(subscription, "check_for_pkg_presence", lambda x: ["test"])
+    @unit_tests.mock(subscription, "get_installed_submgr_pkg_names", lambda x: ["test"])
+    @unit_tests.mock(utils.changed_pkgs_control, "track_installed_pkgs", lambda x: None)
+    def test_install_rhel_subscription_manager(self):
+        subscription.install_rhel_subscription_manager()
+        self.assertTrue("\nPackages installed:\n" in logging.Logger.info.msg)
```
Comment (SpyTec) [inline convert2rhel/unit_tests/subscription_test.py:305]: No assertions are being made about if packages get untracked from the tracked packages in `utils`. `check_for_pkg_presence` could return correct packages but we don't make any assertions that the return value in `rpms_to_untrack` is being used. Same applies to `pkg_names`

If I were to do this I'd expect a test to fail
```python
# rpms_to_untrack = check_for_pkg_presence(pkg_names)
rpms_to_untrack = []
# or 
rpms_to_untrack = ["should-fail"]
```

File: convert2rhel/subscription.py [hunk 8]
```diff
@@ -223,13 +223,17 @@ def remove_original_subscription_manager():
...
     rpms_to_install = [os.path.join(SUBMGR_RPMS_DIR, filename) for filename in os.listdir(SUBMGR_RPMS_DIR)]
 
     if not rpms_to_install:
-        loggerinst.warn("No RPMs found in %s." % SUBMGR_RPMS_DIR)
+        loggerinst.warning("No RPMs found in %s." % SUBMGR_RPMS_DIR)
         return
 
+    pkg_names = get_installed_submgr_pkg_names(rpms_to_install)
+    rpms_to_untrack = check_for_pkg_presence(pkg_names)
+
```
Comment (SpyTec) [inline convert2rhel/subscription.py:235]: Can be moved below the submgr install

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/subscription.py (hunk 8 above)
Comment (r0x0d) [reply convert2rhel/subscription.py:235]: Actually, these calls need to be before the submr install, we need to track any packages that are present on the system before the installation of submr, if, for example, `python3-syspurpose` or `python3-cloud-what` were already present on the system (and may be used by other packages other than submr, we need to untrack them and let those packages be present on the system after the conversion)

File: convert2rhel/unit_tests/subscription_test.py (hunk 7 above)
Comment (r0x0d) [reply convert2rhel/unit_tests/subscription_test.py:305]: You're right! Actually, this reminds me that the `rpms_to_untrack` was not being asserted on the code either!

File: convert2rhel/subscription.py (hunk 3 above)
Comment (r0x0d) [reply convert2rhel/subscription.py:226]: Yeah, but the `loggerinst.info("Installing subscription-manager RPMs.")` is just at the line 236, since below that log is the actual install, I just changed this one because it makes more sense since we are checking for the availability of the rpms first, and then, installing the submr

File: convert2rhel/subscription.py (hunk 8 above)
Comment (SpyTec) [reply convert2rhel/subscription.py:235]: You're right. I'd add a comment instead then so we don't accidentally move it in the future to be closer

File: convert2rhel/subscription.py [hunk 9]
```diff
@@ -242,13 +246,33 @@ def install_rhel_subscription_manager():
...
+    """Returns a list of packages that are already present in the system before
+    installing any packages needed by subscrpition-manager
+
+    :param pkg_names: List of package names
+    """
+    rpms_present = []
+    # Check for already installed packages that subscription-manager uses.
+    # If found any, then add it to a list an consider it as an untracked pkg.
+    for pkg in pkg_names:
+        _, ret_code = utils.run_subprocess("rpm -q %s" % pkg)
```
Comment (SpyTec) [inline convert2rhel/subscription.py:271]:
```suggestion
        _, ret_code = utils.run_subprocess("rpm -q %s" % pkg, print_output=False)
```

File: convert2rhel/subscription.py (hunk 5 above)
Comment (SpyTec) [reply convert2rhel/subscription.py:263]: Reason for this change is because the function is unrelated to subscription-manager and is very generic

File: convert2rhel/subscription.py (hunk 6 above)
Comment (SpyTec) [reply convert2rhel/subscription.py:269]: Reason for this change is because the function is unrelated to subscription-manager and is very generic

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/subscription.py (hunk 8 above)
Comment (r0x0d) [reply convert2rhel/subscription.py:235]: Sure! Good idea. Will add this command right now

**Comment by r0x0d** [comment]
Just a reminder for me (and the reviewers too). 

If this PR get merged before the https://github.com/oamg/convert2rhel/pull/389, then, @abadger and @Andrew-ang9 you guys have to change the `run_subprocess()` calls I've added 

* https://github.com/r0x0d/convert2rhel/blob/untrack-python3-syspurpose/convert2rhel/subscription.py#L271

Otherwise, if this gets merged after the #389, then I have to rebase and modify the `run_subprocess()` call.

File: convert2rhel/unit_tests/subscription_test.py (hunk 7 above)
Comment (r0x0d) [reply convert2rhel/unit_tests/subscription_test.py]: I've updated the test with some assertions, but don't know if they are the right ones. I believe with the assertations I just updated, we can validate that the `rpms_to_untrack` was rightfully called?

File: convert2rhel/subscription.py [hunk 10]
```diff
@@ -244,11 +253,33 @@ def install_rhel_subscription_manager():
...
+def check_for_pkg_presence(pkg_names):
+    """Check that packages are installed on the system.
+
+    :param pkg_names: List of package names
+    """
+    rpms_present = []
+    # Check for already installed packages.
+    # If a package is installed, add it to a list which is returned.
+    for pkg in pkg_names:
+        _, ret_code = utils.run_subprocess("rpm -q %s" % pkg, print_output=False)
```
Comment (abadger) [inline convert2rhel/subscription.py:278]:
```suggestion
        _, ret_code = utils.run_subprocess(["rpm", "-q", pkg], print_output=False)
```

**Review by mkluson — APPROVED** [review]

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        c2r.sendline("y")
+        c2r.expect("The tool allows rollback of any action until this point.")
+        c2r.sendline("n")
+        c2r.expect("Calling command 'subscription-manager unregister'")
+        c2r.expect("System unregistered successfully.")
+
+    # check that packages still are in place
+    assert shell("rpm -qi usermode").returncode == 0
+    assert shell("rpm -qi rhn-setup").returncode == 0
+    if "centos-7" in platform.platform():
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:38]: We should extend the condition for oracle linux here as well.

**Review by kokesak — CHANGES_REQUESTED** [review]
Looks good, Since we are testing on all distros (cent7/8 and ol7/8) I would add the equivalent checks to the OL distros that we have for CentOS.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+def test_check_untrack_pkgs_force(convert2rhel, shell):
+    # Terminate the c2r process forcefully,
+    # so it performs rollback.
+    with convert2rhel(f"-y --no-rpm-va") as c2r:
+        c2r.expect("Username")
+        # Due to inconsistent behaviour of 'Ctrl+c'
+        # use 'Ctrl+d' instead
+        c2r.sendcontrol("d")
+        assert c2r.exitstatus != 0
+    if "centos-8" in platform.platform():
```
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:78]: Sure, will do

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        assert_not_installed(shell, "python3-syspurpose")
+    elif "oracle-7" in platform.platform():
```
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:80]: Okay, good point, thanks!

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 11 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:38]: Will do!

File: convert2rhel/subscription.py [hunk 14]
```diff
@@ -244,11 +253,33 @@ def install_rhel_subscription_manager():
...
+    if rpms_to_untrack and pkg_names:
+        for rpm in rpms_to_untrack:
+            loggerinst.debug("Untracking package: %s" % rpm)
+            pkg_names.remove(rpm)
+
+    loggerinst.debug("Tracking installed packages: %r" % pkg_names)
+    utils.changed_pkgs_control.track_installed_pkgs(pkg_names)
+
+
+def check_for_pkg_presence(pkg_names):
```
Comment (bocekm) [inline convert2rhel/subscription.py:269]: 1. The function name `check_for_pkg_presence` does not imply clearly what it does. I'd rename it to e.g. `filter_installed_pkgs()`.
2. The function would be better suited for `pkghandler.py` rather than this `subscription.py` module.
3. The line `_, ret_code = utils.run_subprocess(["rpm", "-q", pkg], print_output=False)` can be moved to a separate function `is_pkg_installed()`
    - this new `is_pkg_installed()` can be used instead of `pkghandler.get_installed_pkg_objects()` within `subscription.py` where simple true/false would suffice (i.e. no need to call yum/dnf python API to retrieve pkg objects)

**Review by bocekm — CHANGES_REQUESTED** [review]

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:78]: Actually, let me reassess a bit. It seems that `python-syspurpose` was removed in [Oracle Linux 7.2](https://docs.oracle.com/en/operating-systems/oracle-linux/7/relnotes7.9/ol7-pkg-appendix.html) and `python3-syspurpose` was removed in [Oracle Linux 8.2](https://docs.oracle.com/en/operating-systems/oracle-linux/8/relnotes8.2/ol8-pkg-appendix.html#ol8-packages-bin-removed-baseos), so there is no need to extend the condition.

[the remark this answers is not in the document]
Comment (kokesak) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:78]: In that case we can maybe add centos8 base repository (https://vault.centos.org/centos/8/BaseOS/x86_64/os/) to OL8 and install the package `python-3syspurpose` from there. Alternatively we can use just the url as the package will not be updated in the future - [python3-syspurpose](https://vault.centos.org/centos/8/BaseOS/x86_64/os/Packages/python3-syspurpose-1.28.21-3.el8.x86_64.rpm).

It feels like a corner case, but since we are already running the test on all platforms, we can add this I think.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def test_check_untrack_pkgs_graceful(convert2rhel, shell):
+    username = "foo"
+    password = "bar"
+    # Provide c2r with incorrect username and password,
+    # so it fails the registration and performs rollback.
+    with convert2rhel(f"-y --no-rpm-va --username {username} --password {password}") as c2r:
+        assert c2r.exitstatus != 0
+    if "centos-8" in platform.platform():
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:63]: We are calling `platform.platform()` many times in all test cases. I would create one module-level variable that would be then used in all if conditions. We can avoid calling the function multiple times.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        c2r.sendline("y")
+        c2r.expect("Continue with the system conversion?")
+        c2r.sendline("y")
+        c2r.expect("The tool allows rollback of any action until this point.")
+        c2r.sendline("n")
+        c2r.expect("Calling command 'subscription-manager unregister'")
+        c2r.expect("System unregistered successfully.")
+
+    # check that packages still are in place
+    assert shell("rpm -qi usermode").returncode == 0
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:36]: We can use the function `assert_not_installed` here as well earlier test case as well.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    assert shell("rpm -qi rhn-setup").returncode == 0
+    if "centos-7" in platform.platform():
+        assert shell("rpm -qi centos-release").returncode == 0
+    elif "centos-8" in platform.platform():
+        assert shell("rpm -qi centos-linux-release").returncode == 0
+
+
+def assert_not_installed(shell, pkg=""):
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:44]: I would rename the function to `assert_is_installed` instead. I cannot see the reason why there is `not` in the name.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 16 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:36]: Good point

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 17 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:44]: Will do

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 15 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:63]: Yeah, that makes sense

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:78]: You're right, will do the same thing for OL7

File: convert2rhel/subscription.py [hunk 18]
```diff
@@ -223,18 +223,27 @@ def remove_original_subscription_manager():
...
-        loggerinst.warn("No RPMs found in %s." % SUBMGR_RPMS_DIR)
+        loggerinst.warning("No RPMs found in %s." % SUBMGR_RPMS_DIR)
         return
 
+    # These commands have to be here, before the call for installing subscription-manager this is intended, otherwise
+    # `pkghandler.filter_installed_pkgs()` would return every single package that is listed in `rpms_to_install` and
+    # we don't want this to happen. We want only the packages that were here before the instalation of
+    # subscription-manager.
+    pkg_names = get_installed_submgr_pkg_names(rpms_to_install)
+    rpms_to_untrack = pkghandler.filter_installed_pkgs(pkg_names)
```
Comment (bocekm) [inline convert2rhel/subscription.py:238]:
```suggestion
    # These functions have to be called before installation of the subscription-manager packages, otherwise
    # `pkghandler.filter_installed_pkgs()` would return every single package that is listed in `rpms_to_install` and
    # we don't want this to happen. We want to know about the packages that were installed before the installation of
    # subscription-manager.
    pkg_names = get_pkg_names_from_rpm_paths(rpms_to_install)
    pkgs_to_not_track = pkghandler.filter_installed_pkgs(pkg_names)
```

File: convert2rhel/subscription.py [hunk 19]
```diff
@@ -244,11 +253,17 @@ def install_rhel_subscription_manager():
...
+    if rpms_to_untrack and pkg_names:
+        for rpm in rpms_to_untrack:
+            loggerinst.debug("Untracking package: %s" % rpm)
+            pkg_names.remove(rpm)
+
+    loggerinst.debug("Tracking installed packages: %r" % pkg_names)
+    utils.changed_pkgs_control.track_installed_pkgs(pkg_names)
 
 
 def get_installed_submgr_pkg_names(rpm_paths):
```
Comment (bocekm) [inline convert2rhel/subscription.py:269]:
```suggestion
def get_pkg_names_from_rpm_paths(rpm_paths):
```

Please move this function to `pkghandler.py` or `utils.py` as it's generic.

File: convert2rhel/subscription.py [hunk 20]
```diff
@@ -244,11 +253,17 @@ def install_rhel_subscription_manager():
...
+    loggerinst.info("\nPackages installed:\n%s" % "\n".join(rpms_to_install))
+
+    # Untrack any packages that are present in the system before the install
+    if rpms_to_untrack and pkg_names:
+        for rpm in rpms_to_untrack:
+            loggerinst.debug("Untracking package: %s" % rpm)
+            pkg_names.remove(rpm)
+
+    loggerinst.debug("Tracking installed packages: %r" % pkg_names)
+    utils.changed_pkgs_control.track_installed_pkgs(pkg_names)
```
Comment (bocekm) [inline convert2rhel/subscription.py:266]:
```suggestion

    track_installed_submgr_pkgs()
    
def track_installed_submgr_pkgs(installed_pkg_names, pkgs_to_not_track):
    """Tracking newly installed subscription-manager pkgs to be able to remove them during a rollback if needed."""
    pkgs_to_track = []
    for installed_pkg in installed_pkg_names:
        if installed_pkg not in pkgs_to_not_track:
            pkgs_to_track.append(pkg)
        else:
            # Don't track packages that were present on the system before the installation
            loggerinst.debug("Skipping tracking previously installed package: %s" % installed_pkg)

    loggerinst.debug("Tracking installed packages: %r" % pkgs_to_track)
    utils.changed_pkgs_control.track_installed_pkgs(pkgs_to_track)
```

**Review by bocekm — CHANGES_REQUESTED** [review]
Some code readability suggestions. No functionality change requested.

File: convert2rhel/subscription.py (hunk 19 above)
Comment (r0x0d) [reply convert2rhel/subscription.py:269]: Done.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 17 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:44]: :heavy_check_mark:

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 15 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:63]: :heavy_check_mark:

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 16 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:36]: :heavy_check_mark:

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+def test_proper_rhsm_clean_up(shell, convert2rhel):
+    """Test that c2r does not remove usermod, rhn-setup and os-release during rollback.
+    Also checks that the system was successfully unregistered.
+    """
+
+    # Ensure usermode and rhn-setup packages are presented
+    assert shell("yum install -y usermode rhn-setup").returncode == 0
+
+    # run c2r until subscribing the system and then emulate pressing Ctrl + C
```
Comment (danmyway) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:15]: @kokesak To be honest, I'm not sure about this comment saying 'emulate pressing Ctrl + C' since there is no emulation as such.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:80]: :heavy_check_mark:

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 11 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:38]: :heavy_check_mark:

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 21 above)
Comment (kokesak) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:15]: Good catch, it's from the old test. Please remove the ctrl+c part

Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:15]: Sure, thanks!

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:78]: :heavy_check_mark:

**Comment by danmyway** [comment]
Review issues resolved, please review @mkluson, @kokesak

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+
+from envparse import env
+
+
+plt = platform.platform()
+# TODO: in integration test for PR#411. Add 'oracle-logos' and 'centos-logos'
+#  to (ol7pkgs, ol8pkgs) and (cos7pkgs, cos8pkgs) respectively
+#  as these missing packages are a result of a different cause
+#  but still missing after a rollback
+ol7pkgs = ["oracle-release-el7", "usermode", "rhn-setup", "python-syspurpose"]
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:11]: I would the variable so it's in uppercase and words are separated by underscore.
```suggestion
OL_7_PKGS = ["oracle-release-el7", "usermode", "rhn-setup", "python-syspurpose"]
```

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+ol8pkgs = ["oracle-release-el8", "usermode", "rhn-setup", "python3-syspurpose"]
+cos7pkgs = ["centos-release", "usermode", "rhn-setup", "python-syspurpose"]
+cos8pkgs = ["centos-linux-release", "usermode", "rhn-setup", "python3-syspurpose"]
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:14]: The same would apply to these as well.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
+import platform
+
+from envparse import env
+
+
+plt = platform.platform()
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:6]: It's nitch picking but I would rename it to something that would represent the value inside, Eg. something like `system_release`

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+#  but still missing after a rollback
+ol7pkgs = ["oracle-release-el7", "usermode", "rhn-setup", "python-syspurpose"]
+ol8pkgs = ["oracle-release-el8", "usermode", "rhn-setup", "python3-syspurpose"]
+cos7pkgs = ["centos-release", "usermode", "rhn-setup", "python-syspurpose"]
+cos8pkgs = ["centos-linux-release", "usermode", "rhn-setup", "python3-syspurpose"]
+# The packages 'python-syspurpose' and 'python3-syspurpose' were removed in Oracle Linux 7.9
+# and Oracle Linux 8.2 respectively.
+# Installing and checking for these packages (just to make sure)
+# from CentOS repositories to cover all distros consistently.
+ol7repobypass = "http://mirror.centos.org/centos/7/os/x86_64/Packages/python-syspurpose-1.24.42-1.el7.centos.x86_64.rpm"
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:19]:
```suggestion
OL_7_REPO_BYPASS = "http://mirror.centos.org/centos/7/os/x86_64/Packages/python-syspurpose-1.24.42-1.el7.centos.x86_64.rpm"
```

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def define_pkg(shell, osys="", pkgs=None):
+    if osys in plt:
+        for pkg in pkgs:
+            print(f"PREP: Setting up {pkg}")
+            query = shell(f"rpm -q {pkg}")
+            if f"{pkg} is not installed" in query.output:
+                if pkg == "python3-syspurpose" and osys == "oracle-8":
+                    shell(f"yum install -y {ol8repobypass}")
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:32]: I would rather put assert here. In case the URL stops working or there is some error we need to know. If the `shell` command returns non-zero return code the pytest won't report a failure.
```suggestion
                    assert shell(f"yum install -y {ol8repobypass}").returncode == 0
```

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+                elif pkg == "python-syspurpose" and osys == "oracle-7":
+                    shell(f"yum install -y {ol7repobypass}")
```
Comment (kokesak) [inline tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:34]: The same as above

**Review by kokesak — CHANGES_REQUESTED** [review]
Looks good, I would just change mentioned things.

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 22 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py:11]: Sure, good point

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py (hunk 26 above)
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py]: After the disscussion with @kokesak and @mkluson it has been decided to abandon checking for python-syspurpose and python3-syspurpose on Oracle Linux completely as these are not available on convertable versions of OL

**Review by kokesak — CHANGES_REQUESTED** [review]

File: tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+# and Oracle Linux 8.2 respectively.
+
+
+def install_pkg(shell, pkgs=None):
+    for pkg in pkgs:
+        print(f"PREP: Setting up {pkg}")
+        assert shell(f"yum install -y {pkg}").returncode == 0
+
+
+def define_booted_os(shell):
```
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/integration/tier0/test-check-rollback-handling/test_check_rollback_handling.py]: Sure, good point!
Done.

**Comment by danmyway** [comment]
Ready for review.
@kokesak, @mkluson - when you'll have time, please.

**Review by kokesak — APPROVED** [review]
Looks good. Thanks for the good work!

**Comment by danmyway** [comment]
@bocekm, @SpyTec could you please check for the requested changes, so we might merge? Thank you!

**Review by bocekm — APPROVED** [review]

**Review by SpyTec — APPROVED** [review]
